### PR TITLE
Write functions for mosfet.c

### PR DESCRIPTION
**Pull request by wilsonchen02**
Also add main.h, hal library, to mosfet.h

### CODE REVIEW

File: science-dev/Core/Src/mosfet.c [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+ * mosfet.c
+ *
+ *  Created on: Sep 25, 2022
+ *      Author: ASD
+ */
+
+#include "mosfet.h"
+
+MosfetDevice *new_mosfet_device(GPIO_TypeDef *_port, uint8_t _pin) {
+	MosfetDevice new_m_dev = {_port, _pin};
```
Comment (tabiosg) [inline science-dev/Core/Src/mosfet.c:11]: not good. it goes out of scope

File: science-dev/Core/Src/mosfet.c [hunk 2]
```diff
@@ -0,0 +1,35 @@
+/*
+ * mosfet.c
+ *
+ *  Created on: Sep 25, 2022
+ *      Author: ASD
+ */
+
+#include "mosfet.h"
+
+MosfetDevice *new_mosfet_device(GPIO_TypeDef *_port, uint8_t _pin) {
```
Comment (tabiosg) [inline science-dev/Core/Src/mosfet.c:10]: make sure to copy over the requires modifies effects from .h

**Review by tabiosg — CHANGES_REQUESTED** [review]
see comments

File: science-dev/Core/Inc/mosfet.h [hunk 3]
```diff
@@ -1,6 +1,10 @@
 #ifndef MOSFET_H_
```
Comment (tabiosg) [inline science-dev/Core/Inc/mosfet.h:1]: See if you can replace with pragma once. Idk if supported on c though

File: science-dev/Core/Src/mosfet.c [hunk 4]
```diff
@@ -0,0 +1,48 @@
+/*
+ * mosfet.c
+ *
+ *  Created on: Sep 25, 2022
```
Comment (tabiosg) [inline science-dev/Core/Src/mosfet.c:4]: Delete, kinda messy

File: science-dev/Core/Src/mosfet.c [hunk 5]
```diff
@@ -0,0 +1,48 @@
+/*
+ * mosfet.c
+ *
+ *  Created on: Sep 25, 2022
+ *      Author: ASD
+ */
+
+#include "mosfet.h"
+
+// REQUIRES: port is the port and pin is the pin
```
Comment (tabiosg) [inline science-dev/Core/Src/mosfet.c:10]: Fix on .c and .h  it should be _port not port and same with _pin

File: science-dev/Core/Src/mosfet.c [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+	GPIO_TypeDef *temp_port = mosfet_device->port;
+	uint8_t temp_pin = mosfet_device->pin;
+	HAL_GPIO_WritePin(temp_port, temp_pin, GPIO_PIN_RESET);
+}
+
+// REQUIRES: mosfet_device is a Mosfet Device object and state is either 0 or 1
+// MODIFIES: nothing
+// EFFECTS: Sets Mosfet Device to desired state
+void set_mosfet_device_state(MosfetDevice *mosfet_device, uint8_t state) {
+	if(state == 1) {
```
Comment (tabiosg) [inline science-dev/Core/Src/mosfet.c:42]: Can you run using clang format editor, see mrover ros wiki style guide for more info

File: science-dev/Core/Inc/mosfet.h [hunk 7]
```diff
@@ -1,6 +1,10 @@
-#ifndef MOSFET_H_
+#pragma once
 #define MOSFET_H_
```
Comment (tabiosg) [inline science-dev/Core/Inc/mosfet.h:2]: delete

**Review by tabiosg — CHANGES_REQUESTED** [review]

File: science-dev/Core/Src/mosfet.c [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+	uint8_t temp_pin = mosfet_device->pin;
+	HAL_GPIO_WritePin(temp_port, temp_pin, GPIO_PIN_RESET);
+}
+
+// REQUIRES: mosfet_device is a Mosfet Device object and state is either 0 or 1
+// MODIFIES: nothing
+// EFFECTS: Sets Mosfet Device to desired state
+void set_mosfet_device_state(MosfetDevice *mosfet_device, uint8_t state)
+{
+	if (state == 1)
```
Comment (tabiosg) [inline science-dev/Core/Src/mosfet.c:39]: Nah flip the if the statements. Right now it does 1 for on and 0 for everything else. If someone puts 2, it goes off which is kinda weird. Either do if 0 and else OR YOU CAN do if 0 and else if 1.

**Review by tabiosg — CHANGES_REQUESTED** [review]
Nit

File: science-dev/Core/Src/mosfet.c (hunk 8 above)
Comment (tabiosg) [reply science-dev/Core/Src/mosfet.c:39]: Ok honestly just include stdbool.h to make it easy

Comment (wilsonchen02) [reply science-dev/Core/Src/mosfet.c:39]: True

**Review by tabiosg — CHANGES_REQUESTED** [review]
nit

**Review by tabiosg — APPROVED** [review]
Looks good! I will just commit directly since different IOC and too many gross changes lol

**Pull request closed without merging**